src/v/serde/protobuf/tests/parser_test.cc

```diff
@@ -0,0 +1,548 @@
+    EXPECT_TRUE(parse(R"(
+# proto-file: test_messages_edition2023.proto
+# proto-message: TestAllTypesEdition2023
+optional_int32: -1
+)"));
+    EXPECT_TRUE(parse(R"(
+# proto-file: test_messages_edition2023.proto
+# proto-message: TestAllTypesEdition2023
+optional_sint32: -1
+)"));
+    EXPECT_TRUE(parse(R"(
+# proto-file: test_messages_edition2023.proto
+# proto-message: TestAllTypesEdition2023
+optional_sfixed32: -1
+)"));
+}
+
+TEST_F(ProtobufParserFixture, RandomData) {
+    protobuf_test_messages::editions::TestAllTypesEdition2023 kitchen_sink;
+    constexpr size_t size = 4_KiB;
+    constexpr int num_iterations = 10;
+    for (int i = 0; i < num_iterations; ++i) {
+        std::string buf;
+        buf.resize(size);
+        random_generators::fill_buffer_randomchars(buf.data(), buf.size());
+        iobuf b;
+        b.append(buf.data(), buf.size());
+        if (kitchen_sink.ParseFromString(buf)) {
+            EXPECT_NO_THROW(
+              ::serde::pb::parse(std::move(b), *kitchen_sink.descriptor())
+                .get());
+        } else {
+            EXPECT_ANY_THROW(
+              ::serde::pb::parse(std::move(b), *kitchen_sink.descriptor())
+                .get());
+        }
+    }
+    // TODO: Write more tests on invalid data
+}
+
+class ProtobufParserFuzzer : public ProtobufParserFixture {
+public:
+    ProtobufParserFuzzer() { mutator_.Seed(testing::FLAGS_gtest_random_seed); }
+
+    void mutate(pb::Message* msg) { mutator_.Mutate(msg, 3_MiB); }
+
+private:
+    protobuf_mutator::Mutator mutator_;
+};
+
+TEST_F(ProtobufParserFuzzer, AllTypes) {
+    protobuf_test_messages::editions::TestAllTypesEdition2023 kitchen_sink;
+    constexpr int num_iterations = 10;
+    for (int i = 0; i < num_iterations; ++i) {
+        mutate(&kitchen_sink);
+        auto binpb = kitchen_sink.SerializeAsString();
+        iobuf buf;
+        buf.append(binpb.data(), binpb.size());
+        auto parsed = ::serde::pb::parse(
+                        std::move(buf), *kitchen_sink.descriptor())
+                        .get();
+        auto msg = convert_to_protobuf(*parsed, kitchen_sink.descriptor());
+        EXPECT_TRUE(proto_equals(*msg, kitchen_sink));
+    }
+}
+
+TEST_F(ProtobufParserFuzzer, AllTypesMerged) {
+    protobuf_test_messages::editions::TestAllTypesEdition2023 kitchen_sink;
+    constexpr int num_iterations = 10;
+    std::string merged_binpb;
+    for (int i = 0; i < num_iterations; ++i) {
+        mutate(&kitchen_sink);
+        merged_binpb += kitchen_sink.SerializeAsString();
+        kitchen_sink.Clear();
+    }
+    iobuf buf;
+    buf.append(merged_binpb.data(), merged_binpb.size());
+    auto parsed
+      = ::serde::pb::parse(std::move(buf), *kitchen_sink.descriptor()).get();
+    auto msg = convert_to_protobuf(*parsed, kitchen_sink.descriptor());
+    EXPECT_TRUE(kitchen_sink.ParseFromString(merged_binpb));
+    EXPECT_TRUE(proto_equals(*msg, kitchen_sink));
+}
+
+} // namespace
```

Comment:
nit: any reason to have all the tests (not just the utility methods) be in an anonymous namespace?

Reply:
Not really test namespace conflicts are unlikely and probably should be compile errors because the test output will be ambiguous